File: src/modules/algorithm/seed/algorithms/ring-buffer/typescript/array-solution.ts
export class RingBuffer<T> {
  private buffer: (T | undefined)[];
  private head: number;
  private tail: number;
  private count: number;
  private readonly capacity: number;

  constructor(capacity: number) {
    this.buffer = new Array(capacity).fill(undefined);
    this.head = 0;
    this.tail = 0;
    this.count = 0;
    this.capacity = capacity;
  }

  enqueue(item: T): void {
    this.buffer[this.head] = item;
    this.head = (this.head + 1) % this.capacity;
    if (this.count === this.capacity) {
      this.tail = (this.tail + 1) % this.capacity;
    } else {
      this.count++;
    }
  }

  dequeue(): T | undefined {
    if (this.isEmpty()) {
      return undefined;
    }
    const item = this.buffer[this.tail];
    this.buffer[this.tail] = undefined;
    this.tail = (this.tail + 1) % this.capacity;
    this.count--;
    return item;
  }

  isFull(): boolean {
    return this.count === this.capacity;
  }

  isEmpty(): boolean {
    return this.count === 0;
  }

  getCount(): number {
    return this.count;
  }
}

File: src/modules/algorithm/seed/algorithms/ring-buffer/python/array-solution.py
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class RingBuffer(Generic[T]):
    def __init__(self, capacity: int):
        self.buffer: List[Optional[T]] = [None] * capacity
        self.head: int = 0
        self.tail: int = 0
        self.count: int = 0
        self.capacity: int = capacity

    def enqueue(self, item: T) -> None:
        self.buffer[self.head] = item
        self.head = (self.head + 1) % self.capacity
        if self.count == self.capacity:
            self.tail = (self.tail + 1) % self.capacity
        else:
            self.count += 1

    def dequeue(self) -> Optional[T]:
        if self.is_empty():
            return None
        item = self.buffer[self.tail]
        self.buffer[self.tail] = None
        self.tail = (self.tail + 1) % self.capacity
        self.count -= 1
        return item

    def is_full(self) -> bool:
        return self.count == self.capacity

    def is_empty(self) -> bool:
        return self.count == 0

    def get_count(self) -> int:
        return self.count 